miprometheus/problems/video_text_to_class/cog/cog_utils/train_utils.py

```diff
@@ -0,0 +1,290 @@
+  """Pre-process images."""
+  if (vis_type == 'vgg') or (vis_type == 'vgg_pretrain'):
+    in_imgs_ -= np.array([_R_MEAN, _G_MEAN, _B_MEAN], dtype=np.float32)
+  else:
+    in_imgs_ /= 255.
+    in_imgs_ -= np.mean(in_imgs_)
+
+  return in_imgs_
+
+
+def tasks_to_rules(tasks):
+  """Generate in_rule and seq_length arrays.
+
+  Args:
+    tasks: a list of tg.Task instances or string rules, length is batch_size.
+  """
+  batch_size = len(tasks)
+  in_rule = np.zeros((const.MAXSEQLENGTH, batch_size), dtype=np.int64)
+  seq_length = np.zeros((batch_size,), dtype=np.int64)
+  for i_task, task in enumerate(tasks):
+    word_list = re.findall(r"[\w']+|[.,!?;]", str(task))
+    seq_length[i_task] = len(word_list)
+    for i_word, word in enumerate(word_list):
+      in_rule[i_word, i_task] = const.INPUTVOCABULARY.index(word)
+  return in_rule, seq_length
+
+
+def set_outputs_from_tasks(n_epoch, tasks, objsets,
+                           out_pnt_xy, out_word,
+                           mask_pnt, mask_word):
+  j = 0
+  for epoch_now in range(n_epoch):
+    for task, objset in zip(tasks, objsets):
+      target = task(objset, epoch_now)
+      if target is const.INVALID:
+        # For invalid target, no loss is used. Everything remains zero.
+        pass
+      elif isinstance(target, sg.Loc):
+        # minimize point loss
+        out_pnt_xy[j, :] = target.value
+        mask_pnt[j] = 1.
+      elif isinstance(target, bool) or isinstance(target, sg.Attribute):
+        if isinstance(target, bool):
+          target = 'true' if target else 'false'
+        else:
+          target = target.value
+        # For boolean target, only minimize word loss
+        out_word[j] = const.OUTPUTVOCABULARY.index(target)
+        mask_word[j] = 1.
+      else:
+        raise TypeError('Unknown target type.')
+      j += 1
+
+
+def set_outputs_from_targets(n_epoch, objsets,
+                             out_pnt_xy, out_word,
+                             mask_pnt, mask_word):
+  j = 0
+  for epoch_now in range(n_epoch):
+    for objset in objsets:
+      target = objset.targets[epoch_now]
+      if target == 'invalid':
+        # For invalid target, no loss is used. Everything remains zero.
+        pass
+      elif isinstance(target, (list, tuple)):
+        assert len(target) == 2, "Expected 2-D target. Got " + str(target)
+        # minimize point loss
+        out_pnt_xy[j, :] = target
+        mask_pnt[j] = 1.
+      elif isinstance(target, string_types):
+        out_word[j] = const.OUTPUTVOCABULARY.index(target)
+        mask_word[j] = 1.
+      else:
+        raise TypeError('Unknown target type: %s %s' % (type(target), target))
+      j += 1
+
+
+def generate_batch(tasks,
+                   n_epoch=30,
+                   img_size=224,
+                   objsets=None,
+                   n_distractor=1,
+                   average_memory_span=2):
+  """Generate a batch of trials.
+
+  Return numpy arrays to feed the tensorflow placeholders.
```

Comment:
That part of documentation seems to be incompatible, I am not sure whether Sphinx/ReadTheDocs will be able to parse it properly. @vmarois can you please double check it?